app/uk/gov/hmrc/transitmovementsrouter/connectors/EISConnector.scala

```diff
@@ -157,7 +161,30 @@ class EISConnectorImpl(
                 } else s"Response status: ${error.statusCode}"
 
               logger.warn(logMessage + status)
-              Left(RoutingError.Upstream(UpstreamErrorResponse(s"Request Error: Routing to $code returned status code ${error.statusCode}", error.statusCode)))
+
+              error.statusCode match {
+                case ErrorCode.Forbidden.statusCode | ErrorCode.InternalServerError.statusCode =>
+                  val lrnMatch = lrnRegexPattern.findFirstMatchIn(error.message)
+                  lrnMatch match {
+                    case Some(message) =>
+                      val lrn: LocalReferenceNumber =
+                        Some(message)
+                          .map(_.group(1))
+                          .map(LocalReferenceNumber.apply)
+                          .get
+                      Left(RoutingError.DuplicateLRNError(s"LRN ${lrn.value} has previously been used and cannot be reused", ErrorCode.Conflict, lrn))
+                    case None =>
+                      Left(
+                        RoutingError.Upstream(
+                          UpstreamErrorResponse(s"Request Error: Routing to $code returned status code ${error.statusCode}", error.statusCode)
+                        )
+                      )
+                  }
```

Comment:
You could reduce this down to 

```suggestion
                  message match {
                    case lrnRegexPattern(lrn) =>
                      Left(RoutingError.DuplicateLRNError(s"LRN $lrn has previously been used and cannot be reused", ErrorCode.Conflict, LocalReferenceNumber(lrn)))
                    case None =>
                      Left(
                        RoutingError.Upstream(
                          UpstreamErrorResponse(s"Request Error: Routing to $code returned status code ${error.statusCode}", error.statusCode)
                        )
                      )
                  }
```

See https://docs.scala-lang.org/tour/regular-expression-patterns.html -- it's quite a cool (in my opinion!) use of pattern matching in Scala. `case lrnRegexPattern(lrn)` matches the string, and extracts the first group as `lrn`.

Reply:
Yes @drnaylor-oc totally agree, this is how I first implemented. But below is the response from eis-stub for the _error.message_ attribute and it did not extracts the first group as `lrn`. 

`POST of 'http://localhost:9476/transit-movements-eis-stub/movements/messages/gb' returned 403. Response body: '{"code":"FORBIDDEN","message":"Error in request: The supplied LRN: LRN12345678 has already been used by submitter: messageSender"}'`

So had to tweak it accordingly. Welcome for any other suggestions.

Reply:
Ah, yeah, because of that extra stuff the error puts around it.

I just wrote a quick scala-cli script that does work:

```scala
//> using scala "2.13"

val pattern = ".*The supplied LRN: ([a-zA-Z0-9]+) has already been used by submitter: ([a-zA-Z0-9]+).*".r

val text = """POST of 'http://localhost:9476/transit-movements-eis-stub/movements/messages/gb' returned 403. Response body: '{"code":"FORBIDDEN","message":"Error in request: The supplied LRN: LRN12345678 has already been used by submitter: messageSender"}'"""

text match {
    case pattern(t, _) => println(s"This worked: LRN $t")
    case _ => println("This didn't work")
}
```

Two things:

* We need to surround the regex with `.*` -- so it'll do a full match on the entire string
* The unapply needs two parameters as there are two groups in your regex (which I didn't notice!)

Reply:
Ah yes I could have updated the regex, didn't realize it. Thank you @drnaylor-oc will update it now.